src/Lucene.Net/Util/UnmanagedStringArray.cs

```diff
@@ -0,0 +1,185 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+
+namespace Lucene.Net.Util
+{
+    public unsafe class UnmanagedStringArray
+    {
+        private class Segment : IDisposable
+        {
+            public static long SegmentNumber;
+
+            public readonly int Size;
+            public readonly long Number;
+
+            public byte* Start;
+            public byte* CurrentPosition => Start + Used;
+            public int Free => Size - Used;
+            public int Used;
+
+            public Segment(int size)
+            {
+                Start = (byte*) Marshal.AllocHGlobal(size);
+                Used = 0;
+                Size = size;
+                Number = Interlocked.Increment(ref SegmentNumber);
+            }
+
+            public void Add(ushort size, out byte* position)
+            {
+                position = CurrentPosition;
+                *(ushort*) CurrentPosition = size;
+
+                Used += sizeof(short) + size;
+            }
+
+            public void Dispose()
+            {
+                GC.SuppressFinalize(this);
+                if (Start != null)
+                    Marshal.FreeHGlobal((IntPtr) Start);
+                Start = null;
+            }
+
+            ~Segment()
+            {
+                Dispose();
+            }
+        }
+
+        public struct UnmanagedString : IComparable
+        {
+            public byte* Start;
+            public long Owner;
```

Comment:
Why do it this way? To avoid forcing a reference and the GC cost of that?

Reply:
I assume that this is for debug purposes only? If so, we want to drop this. It is actually too expensive for us (doubles the cost of this struct).